### PR TITLE
Enterprise feature: cobranding watermark

### PR DESCRIPTION
**Pull request by kopijunkie**
# Problem
Some Enterprise customers want to have the footer watermark display as a co-brand instead of the default Onfido only branding

Screenshot 2020-06-25 at 14 37 17

# Solution
* Added co-branded watermark on all SDK screens for mobile and tablet devices 
    * note two versions for light/dark backgrounds
* Co-branding is displayed as a text option in SDK initialisation config
    * if "XXX powered by Onfido" exceeds the allowed space (with paddings on both sides of the screen), then "XXX" text is truncated
* Feature only visible if is enabled for the customer, otherwise will still see default watermark
the enterprise hide logo feature should override this feature
* The enterprise feature to hide footer watermark overrides this feature

## Checklist
_put `n/a` if item is not relevant to PR changes_

- [x] Has the CHANGELOG been updated?
- [x] Has the README been updated?
- [ ] Has the CONTRIBUTING doc been updated?
- [n/a] Has the RELEASE_GUIDELINES been updated?
- [n/a] Has the MIGRATION doc been updated for any MAJOR breaking changes?
- [n/a] Has the MIGRATION doc been updated for any MINOR breaking changes, including any translation strings or keys changes?
- [x] Have any new automated tests been implemented or the existing ones changed?
- [n/a] Have any new manual tests been written down or the existing ones changed?
- [n/a] Have any new strings been translated or the existing ones changed?

### CODE REVIEW

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-1057-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-1057-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://1057-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-1057-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-1057-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-1057-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-1057-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://1057-pr-onfido-sdk-ui-onfido.surge.sh

File: src/components/Router/StepsRouter.js [hunk 1]
```diff
@@ -20,19 +20,33 @@ class StepsRouter extends Component {
...
     const stepId = `onfido-step${this.props.step}`  // to trigger update in NavigationBar on step change
-    // This prevents the logo from appearing late
-    const hideLogoLogic = mobileFlow ? hideOnfidoLogo : globalUserOptions.enterpriseFeatures?.hideOnfidoLogo && hideOnfidoLogo
+    // This prevents the logo, cobrand UI elements from appearing late
+    const hideLogoLogic = mobileFlow ?
+      hideOnfidoLogo :
+      globalUserOptions.enterpriseFeatures?.hideOnfidoLogo && hideOnfidoLogo
+    const cobrandLogic = mobileFlow ?
+      cobrand :
+      globalUserOptions.enterpriseFeatures?.cobrand
```
Comment (eraxyz) [inline src/components/Router/StepsRouter.js:42]: Should this be `globalUserOptions.enterpriseFeatures?.cobrand && cobrand`? Or will that cause it to appear after the rest of the page has loaded? The problem with this is it allows users to use the cobrand feature even if they don't have the feature enabled on their account.

File: src/demo/demoUtils.js [hunk 2]
```diff
@@ -52,25 +52,32 @@ export const getInitSdkOptions = () => {
...
-  const smsNumberCountryCode = queryParamToValueString.countryCode
-    ? { smsNumberCountryCode: queryParamToValueString.countryCode }
-    : {}
+  const smsNumberCountryCode = queryParamToValueString.countryCode ?
+    { smsNumberCountryCode: queryParamToValueString.countryCode } :
+    {}
 
   const hideOnfidoLogo = queryParamToValueString.hideOnfidoLogo === 'true'
+  const cobrand = queryParamToValueString.showCobrand === 'true' ?
+    { text: "Planet Express, Incorporated" } :
```
Comment (eraxyz) [inline src/demo/demoUtils.js:61]: 👍

File: src/components/Router/StepsRouter.js (hunk 1 above)
Comment (kopijunkie) [reply src/components/Router/StepsRouter.js:42]: I initially had it as `globalUserOptions.enterpriseFeatures?.cobrand && cobrand` but that had the side effect of causing it to appear only after the rest of the page has loaded so the default logo is still visible for a split second. Any suggestions for how to get around this?

File: src/components/App/index.js [hunk 3]
```diff
@@ -103,9 +104,20 @@ class ModalApp extends Component {
         this.onInvalidJWT('hideOnfidoLogo feature not enabled for this account.')
       }
 
-    } else if (!options.mobileFlow && token && token !== prevToken) {
+    } else if (!options.mobileFlow && isNewToken) {
       this.props.actions.hideOnfidoLogo(false)
     }
+    if (options.enterpriseFeatures?.cobrand && !isHideOnfidoLogoEnabled && isNewToken) {
```
Comment (stefaniacardenas) [inline src/components/App/index.js:110]: The `prepareInitialStore` function is 51 lines and it's full of if else statement. Could we split it into smaller functions to make it more readable?

File: src/components/App/index.js [hunk 4]
```diff
@@ -103,9 +104,20 @@ class ModalApp extends Component {
...
+      const validEnterpriseFeatures = getEnterpriseFeaturesFromJWT(token)
```
Comment (stefaniacardenas) [inline src/components/App/index.js:111]: Should we store these values in the redux store so we don't have to call this function to parse the token every time? Or you can just call it at the beginning of the function and delete it from line 97 too.

File: src/components/App/index.js [hunk 5]
```diff
@@ -103,9 +104,20 @@ class ModalApp extends Component {
...
+      if (validEnterpriseFeatures?.cobrand) {
+        this.props.actions.showCobranding(options.enterpriseFeatures.cobrand)
+      } else {
+        this.props.actions.showCobranding(null)
+        const errorMessage = 'EnterpriseFeatureNotEnabledException: Enterprise feature cobrand not enabled for this account.'
+        Tracker.trackException(errorMessage)
+        this.onInvalidJWT(errorMessage)
```
Comment (stefaniacardenas) [inline src/components/App/index.js:118]: If the JWT is invalid, by the time this logic is run, the host app would have already received the onError error callback so I don't see this why we need to send it again. Maybe we need return enterprise features errors in the onError callback...

File: src/components/App/index.js (hunk 3 above)
Comment (kopijunkie) [reply src/components/App/index.js:110]: Good point! Will work on splitting it up to be more manageable & readable.

File: src/components/App/index.js (hunk 4 above)
Comment (kopijunkie) [reply src/components/App/index.js:111]: I think it's only used here, so we could call it at the beginning of the function to start with.

File: src/components/App/index.js (hunk 5 above)
Comment (kopijunkie) [reply src/components/App/index.js]: I've split the enterprise feature errors into a separate error handler. This should make it clearer which type of JWT error is being called.

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-1057-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-1057-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-1057-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-1057-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://1057-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://1057-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-1057-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-1057-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://1057-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-1057-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-1057-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://1057-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-1057-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-1057-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-1057-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-1057-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://1057-pr-onfido-sdk-ui-onfido.surge.sh

**Pull request merged**